geode-core/src/main/java/org/apache/geode/cache/client/internal/QueryOp.java

```diff
@@ -120,7 +122,11 @@ protected Object processResponse(Message msg) throws Exception {
           queryResult = resultPart.getObject();
         } catch (Exception e) {
           String s = "While deserializing " + getOpName() + " result";
-          exceptionRef[0] = new SerializationException(s, e);
+          if (e instanceof PdxSerializationException) {
+            exceptionRef[0] = new IOException(s, e);
```

Comment:
Can you please explain the reasoning behind converting a `PdxSerializationException` to an `IOException`?

It is a bit puzzling looking at this code (without really understanding the calling context), why we would convert an exception with "serialization" in its name, to an `IOException` but at the same time convert all other exceptions to `SerializationException`.

I assume there is some sort of contract in play here. But there is no method-level comment describing that contract. Nor does the base class `AbstractOp` explain that contract.

It would be very helpful if that reasoning was captured in a method-level doc comment. If the reasoning applies only to `QueryOp` it could go on this method, but I suspect the reasoning actually applies at the level of `AbstractOp`, in which case, the documentation could go there.

Reply:
The reason is explained in GEODE-9346 ticket. 

In short, IOException will allow the client to retry next server in the connection pool until exhaust all the servers (so it will not retry forever). 

Since the byte array of the pdxInstance is always the same at the server and other clients can deserialize it correctly, even this client can get the pdxInstance before and after the exception, retry is a feasible solution. 

I will add this comment in the code too.

Reply:
Thank you for the extra comment in `processResponse()` @gesterzhou! I now understand that by throwing an `IOException` from any `AbstractOp.processResponse()` (like `QueryOp.processResponse()`) it will cause retrying by `OpExecutorImpl.execute()` and `OpExecutorImpl.handleException()`.

I understand your reasoning, that if those methods in `OpExecutorImpl` are retrying (on a single server, and across servers in the pool), and if _any_ server returns valid chunks (within the retry window), the client will eventually succeed.

What I do not understand, from this PR, and what I do not see represented in tests, is the mechanism by which the corruption of chunks occurs in the first place. You theorize that when a client communicates with a server that is crashing, that we might see corrupt chunks. If that's true, I'd ideally like test code to show that. And if that is infeasible, I'd like a more in-depth explanation of how that could happen.

Without that understanding, I am left wondering: what are the _other_ ways (besides your theorized crashing-server-sends-corrupt-chunk) that the second part of a chunk could be corrupted. And are there ways it could be corrupted on _all_ the servers—not just a single, crashing one.

And if a chunk were corrupted on all the servers, this PR would cause the client to retry across all servers, a total of `servers * retries` tries. If I adopt your theory then perhaps this is not deal. But if I do not adopt your theory, I have a fear that a corrupt chunk could appear on all servers frequently and this could cause a lot of extra client-server traffic.

If I believed your theory, I'd wonder: how (and where, in the code) can we end up with a chunk that is corrupt. That is where the bug seems to lie. The current PR papers over that problem by masking it as an `IOException`. It makes me think there was a deeper `IOException` we failed to handle properly.

Reply:
It won't corrupted on all servers. All the servers are hosting the same byte array. Other clients get get it correctly. Even the same client can get it correctly later. And the socket problem only happens occasionally, which is proven by the customer. 

The philosophy of the retry is to handle those occasionally network issue. Geode have retry logic for all client to server operation.

Reply:
The problem only happen once every 2 months.

Reply:
If there was a bug (a different bug than GEODE-9346) in the code that serialized the PDX object then every copy would be corrupt right? I wonder how likely that sort of bug is and I wonder what the impact would be with the solution in this PR. In that case every server would be retried.

Another concern I have is I wonder if the fix in the PR might be insufficient to fix the problem when other kinds of serialization besides PDX are used (for domain objects).

In the PR scenario, a `PdxSerializationException` is encountered when deserializing domain objects returned by a query. I assume this is because the application happens to be doing this: https://geode.apache.org/docs/guide/12/developing/data_serialization/use_pdx_high_level_steps.html

If instead, the application was using standard Java serialization like https://geode.apache.org/docs/guide/12/developing/data_serialization/java_serialization.html would we expect to see some subclass of `ObjectStreamException` (which happens to be a subclass of `IOException`) instead? see reference here: https://facingissuesonit.com/2019/08/29/java-serialization-exception-handling/

If we believed that we might see these different exceptions for Java serializables then that might suggest that we should catch those too.

It seems to me that this bug (GEODE-9346) is caused by the `Message` class failing to do its job. That class is responsible for deserializing messages from sockets. `readHeaderAndBody()` is responsible for reading complete, correct messages. If the body contains a corrupt payload, doesn't that indicate we have a bug down in `readPayloadFields()`? The `Message` class uses `ByteBuffers`. Are those buffers shared? Because this this seems like the kind of problem that happens when shared buffers are corrupted.

Reply:
There're 4 questions. I will answer them one by one:

(1) If there's a real PdxSerialization corruption happened at server side caused by application (it did not happen so far though. Customer has verified all their server pdx types and they kept verifying from time to time). It's user's responsibility to fix their pdx serialzation issue in their application code. Our responsibility is to tell the user that their pdx type is wrong. That's why I add a dunit test to prove that in such a case, the client will fail after retry. My retry logic won't retry forever. It will try all servers then return with exception. You concern if we should fail earlier? I think we are fine to take a few more seconds to double check, because customer has to fix their pdx type in such a case, that will take hours or even days. 

(2) Customer is following the document you mentioned to use pdx type in correct way. That's why the client query correctly even before and after the occasional exception. If they are using the wrong way, they will encounter problem everyday. 

The application is querying the pdx object's byte array from server, then deserialized it at client. This is our default behavior. We offer another non-default usage is: set --read-serialized=true, then the client will not deserialize the byte array and wrap it as a PdxInstance object. It will delay the deserialization into their application logic field-by-field. But that's not our default setting.  

(3) If application is using standard Java serialization, then it will not fail with PdxSerializationException. It will not fail with ObjectStreamException either. It will fail with SerialzationException, which is handled in "else" (see my diff). I purposely did not change SerialzationException into IOException. The reason is as customer support ticket, we only fix what the issue is. If someday we encounter SerialzationException caused by socket corruption, we can fix it later, but not right now. If we concern about SerialzationException at client side should be retried, we can log a new geode ticket. 

(4) I suspected some socket corruption happened, which caused the Message packed wrong byte array. From log, every reproduce of the PdxSerializationException failed in different byte. 
But since it's very hard to reproduce this issue, we have no clue where the trigger is. Maybe in readHeaderAndBody() or readPayloadFields(). Maybe not. Again, if we think these codes have potential issues, we can log a new geode ticket.